### PR TITLE
Pageserver disaster recovery RFC

### PR DESCRIPTION
**Pull request by arpad-m**
Enable the pageserver to recover from data corruption events by implementing
a feature to re-apply historic WAL records in parallel to the already occurring
WAL replay.

The feature is outside of the user-visible backup and history story, and only
serves as a second-level backup for the case that there is a bug in the
pageservers that corrupted the served pages.

The RFC proposes an API endpoint for the pageserver that can be triggered
to create a new copy of a timeline, applying the WAL of a different
timeline, either from the start, right after initdb, or from a historic point.

### CODE REVIEW

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+limits and billing we apply to existing timelines.
+
+## Proposed implementation
+
+The first problem to keep in mind is the reproducability of `initdb`.
+So an initial step would be to upload `initdb` snapshots to S3,
+maybe deduplicating them by hash, for a given storage/postgres version
+combination.
+
+After that, we'd have the endpoint spawn a background process which
```
Comment (jcsp) [inline docs/rfcs/027-pageserver-disaster-recovery.md:133]: It would be helpful to define the boundary between the new code in the background process, and existing code.

e.g. one possible implementation would be to use the existing Timeline class, and the only difference with a recovery timeline is that it consumes the WAL for a different timeline ID than its own ID.  Another possible implementation would be to write a totally fresh type that just implements WAL replay to S3 objects.

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+## Non Goals
+
+At the danger of being repetitive, the main goal of this feature is to be a
+backup method, so reliability is very important. This implies that other
+aspects like performance or space reduction are less important.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Most changes would happen to the pageservers.
+For the higher level features, maybe other components like the console would
```
Comment (jcsp) [inline docs/rfcs/027-pageserver-disaster-recovery.md:120]: I think the main thing to define about console's involvement is how it can make use of the recovered timelines.

e.g. we might call directly to a pageserver to initiate a recovery to some timeline ID, but what happens when the recovery is done and the user wants to start an endpoint targeting this new timeline ID?   Somehow the console needs to learn about this timeline ID that was created out of band (or, the console needs to pick the ID and initiate the recovery).

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+be involved.
+
+We need to make sure that the shadow timelines are not subject to the usual
```
Comment (jcsp) [inline docs/rfcs/027-pageserver-disaster-recovery.md:123]: Good point -- can you be more specific about where in the stack we would track the status of a timeline as either a recovery timeline (free) or a regular timeline (paid)?  It's not obvious to me if this is something that would be internal to the pageserver, or something that the console needs to track.

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
```
Comment (jcsp) [inline docs/rfcs/027-pageserver-disaster-recovery.md:96]: It would be good to define:
- the state of the system is when this API call returns (e.g. does the timeline "exist" from the point of view of a "GET /tenant/foo/timeline/bar" once we've called this API, or only later when the recovery completes?)
-  how the caller can check for progress/completion.
- after recovery is complete, does the timeline automatically transition to being a "normal" timeline that is readable by endpoints?  Or is there some extra step needed later?

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+## Proposed implementation
+
+The first problem to keep in mind is the reproducability of `initdb`.
+So an initial step would be to upload `initdb` snapshots to S3,
+maybe deduplicating them by hash, for a given storage/postgres version
+combination.
+
+After that, we'd have the endpoint spawn a background process which
+performs the replay of the WAL to that new timeline.
+
```
Comment (jcsp) [inline docs/rfcs/027-pageserver-disaster-recovery.md:135]: After recovery, if we want to use the recovered timeline from an endpoint for writes as well as reads, how will that work?

Thinking of when we have original timeline A, and we run recovery timeline B, targeting the "WAL A".  If we then attach an endpoint to timeline B, and it wants to do some writes, does timeline B then need to consume from a new "WAL B"?  Perhaps the idea is that after recovery the B timeline will automatically switch from consuming WAL A to WAL B, but it would be good to make that part explicit.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 1 above)
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:133]: I don't really know the internal architecture of the pageserver well enough to answer this question weighing about different implementation tradeoffs. I would therefore love to defer this question to the implementation phase.

Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:133]: Thinking about it more, I think that it would be better to stay as close as possible to a timeline that is just a bit out of sync with the WAL. The major difference to normal operations would be the start, as in, we'd have to do something special there, and that the timeline uses a different timeline's WAL records until it reaches a specific lsn, at which it plays back its own WAL records.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 4 above)
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:96]: > the state of the system is when this API call returns

I suppose we want the API call to immediately return the timeline ID and then give the caller some means to track the status of the process.

We definitely will not be able to serve get requests for "current" lsns, but for lsns at past moments we would probably be able to. I think there are analogies here with asking for future lsns where WAL replay has not caught up yet? I'd model the process, also alluded to [in my reply above](https://github.com/neondatabase/neon/pull/5248#discussion_r1321832429):

* Start is special, takes reproducible initdb snapshot (cc https://github.com/neondatabase/neon/issues/2592). API call for new timelines also probably returns after initdb ran? idk.
* timeline is created with "wal from other timeline until lsn no `target`" status, probably saved somewhere in the metadata?
* pageserver does its thing where it replays the WAL to make pages
* timeline reaches the `target` lsn: now it changes its status to turn off the "wal from other timeline until lsn `target`" flag, and listens to WAL input for its own timeline ID.

> does the timeline "exist" from the point of view of a "GET /tenant/foo/timeline/bar" once we've called this API, or only later when the recovery completes
>
> how the caller can check for progress/completion.

The console has some kind of "operations" concept, but I don't think we need such in the pageserver: we can just return whatever lsn the process is currently at, and the lsn we want to arrive at is known to the caller, so they can derive a nice percent value in terms of progress from that.

> after recovery is complete, does the timeline automatically transition to being a "normal" timeline that is readable by endpoints? Or is there some extra step needed later?

I think we can just serve all read calls up to the lsn we have replayed.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 5 above)
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:135]: My comment https://github.com/neondatabase/neon/pull/5248#discussion_r1321851512 is along the same lines. Will make it explicit in the RFC.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 4 above)
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:96]: > If our recovery code is a specialized instantiation of a Timeline object with a new ID, then a bug in the pageserver might occur just the same during recovery as it did to begin with. I think that's okay, we just need to be consistent in how we think/talk about this feature

Good point. The goal is not to reimplement a simpler clone of the entire stack, say without compaction, but it's to use a non-buggy pageserver to undo the mess created by a buggy pageserver. Or even if there was no bug, cosmic rays are a thing (so are CT [bit](https://news.ycombinator.com/item?id=27728287) [flips](https://news.ycombinator.com/item?id=35883910)).

I will reword that part to make it clearer :+1: .

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 3 above)
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:123]: I need to research on how the billing infrastructure works. If there is billing specific APIs, push or pull based, then the answer is easy: just pretend the timeline doesn't exist. If billing uses the normal APIs, I wouldn't really like completely hiding the timeline: it's still *three*, and then adding some flag would make more sense, or cross referencing with the timelines that have actual branches/projects associated with them in the console.

What I am trying to say with this is that even though these timelines are in the tenant, they should not have negative consequences on the user, especially if we execute this on random tenants.

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+## Summary
+
+Enable the pageserver to recover from data corruption events by implementing
+a feature to re-apply historic WAL records in parallel to the already occurring
+WAL replay.
+
+The feature is outside of the user-visible backup and history story, and only
+serves as a second-level backup for the case that there is a bug in the
+pageservers that corrupted the served pages.
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:14]: I'm not sure what FPIs are, but in general it is a really hard problem if the pageserver introduced some corrupt reads and then application code, or some other higher level component, like postgres, made some writes based on that corrupt value. I think this should be out of scope for this RFC and is generally quite hard.

[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:14]: i will add this to the non-goals section :+1: . note that corrupting data is not enough, by which I mean corruption of the pageserver internal datastructures. Thee corruption also has to make it to the WAL. obviously an approach that plays back the WAL verbatim will not be enough in that instance.

[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:14]: @MMeent I have added a section. Do you like it?

**Review by petuhovskiy — COMMENTED** [review]
> The feature is outside of the user-visible backup and history story, and only
serves as a second-level backup

+1. I think this feature is about engineer tooling. We (engineers) will use it just like any other available tool, so we should think about the value this tool will provide and in what cases it will be useful.

Therefore, I'd like to see more examples of "disasters" or "debug" usecases that we had or may have, with the explanation of how to use this feature for the purpose.


I looked at mentions of https://github.com/neondatabase/cloud/wiki/Storage:-Recovery-from-WAL in the Slack to get a list of some known examples (you can search `Recovery-from-WAL` in the Slack to get more).

- https://neondb.slack.com/archives/C033RQ5SPDH/p1661408688463559?thread_ts=1661356205.064629&cid=C033RQ5SPDH
    - Konstantin made a point about corruption investigation. For me the idea here that it can be useful to replay WAL from the start for some kind of investigation, and not necessary on the same pageserver (for example create a tenant clone on the staging pageserver, from the production WAL).
- https://neondb.slack.com/archives/C033RQ5SPDH/p1664890530867019
    - It looks like we had some bad S3 intervention on staging, which corrupted/deleted pageserver S3 data. In this "disaster recovery" case we should recreate all the timelines for a given tenant, ignoring the S3 content. Currently there are no good tools for that (timeline hierarchy can be complicated to easily recreate it in the pageserver).
    - Note that we can use the same logic in case of storage format change, to replay all the WAL to generate data in the new format on the pageserver.
- https://neondb.slack.com/archives/C03F5SM1N02/p1665406677551819?thread_ts=1665362497.649899&cid=C03F5SM1N02
    - Corruption due to no free space on disk.
- https://neondb.slack.com/archives/C03H1K0PGKH/p1666184694221939
    - There were some kind of bug or cleanup on staging, after which some projects got stuck. This message has some details about how to get a legacy initdb archive.
- https://neondb.slack.com/archives/C033RQ5SPDH/p1692280956166909
    - We needed to increase retention period for the customer. The idea here is to replay the WAL from the beginning with increased PITR, then customer can make a branch to recover the data.


These examples made me think of the following approach to the disaster recovery:
- let's say we have a corrupted tenant on pageserver ps1
- we don't touch it, because we want to keep it for investigation and to prevent compute downtime
- instead, we will recreate a tenant on another pageserver ps2
- the tenant is recreated from scratch, that means that ps2 doesn't use S3
- but it gets the WAL from the same safekeepers, so in the end it will replay all the WAL from the beginning
- API has an easy way to recreate the full timeline hirerachy and we can pass a custom PITR interval
- after replay is finished, we can just switch project in control-plane from ps1 to ps2 by updating pageserver_id

What do you think about this approach?

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:84]: the use case I can think of is a data corruption bug of the pageserver within PITR, and it being introduced by a certain deploy and we know the date it happened at, from which we can derive the lsn.

probably at the start it's not going to be needed, I agree.

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
+The API call returns immediately after creation of the new timeline, and
+doesn't wait for the timeline's WAL replay to finish. The API call returns
+the id of the newly created timeline. The new timeline will appear in the
+usual listings and its lsn will increase until WAL replay has finished.
+As the caller knows the targeted lsn goal, this allows them to be informed
+about progress. There is no push notification for when the process has
+finished, any followup actions need to first poll for completion.
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:103]: > I don't fully understand this API. Does it work just like the normal branching, but replays the WAL from the beginning?

Yes.

> Is there a big difference to creating a timeline with ancestor_timeline_id = :timeline_id, ancestor_start_lsn = :to_lsn?

You wrote the difference above, which is that we replay the WAL instead of using the pageserver's internal datastructures to perform the branching. ideally, yes, the end result would be quite the same but the idea is that this recovery process is done when there has been a corruption.

We have discussed at the storage offsite whether the timeline should have the original timeline as its anchestor or not. Ideally we'd delete the original timeline however, otherwise it's staying around as a "ghost", eating up data. But then you'll need to be able to handle deleted timelines...

personally I think it would be best to not link to the original timeline as the anchestor but instead add the capability for the timeline to be the new "root" and instead handle things from there.

For recovering non-main branches

File: docs/rfcs/027-pageserver-disaster-recovery.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply docs/rfcs/027-pageserver-disaster-recovery.md:93]: it's just the way of specifying the ID of the new timeline. alternatively, one could rename `target` to `source` and make `:timeline_id` be the ID of the newly created timeline. I'm fine with either, please tell me which one you like more.

But yeah, the pageserver will not expect any WAL pertaining to lsns from the specified range to come from safekeepers for the newly created timeline (target). Instead, it takes the WAL from the source timeline. Also from the safekeepers.

**Comment by arpad-m** [comment]
Thanks @petuhovskiy for going through the past events on slack and summarizing them! This is very helpful.

> instead, we will recreate a tenant on another pageserver ps2
> [...]
> the tenant is recreated from scratch, that means that ps2 doesn't use S3
> [...]
> API has an easy way to recreate the full timeline hirerachy and we can pass a custom PITR interval
> after replay is finished, we can just switch project in control-plane from ps1 to ps2 by updating pageserver_id

i suppose your idea implies that the tenant ID is not being changed? I agree that S3 is the source of potential conflicts, but I don't think that recreating a tenant from scratch will be enough for avoiding the S3 issue.

Eventually we want to upload to S3 and then there is the potential conflicts. We also want to design the feature in a way that we can use it to create a shadow replay of the WAL without affecting computes at all, so we cannot just delete or rename the entire S3 prefix for that tenant. We could introduce a new "dimension" for the data, like @jcsp introduced via his generations changes (generations are not suited from me looking at them, as they still keep the invariant that the data for a given LSN never changes between generations). IMO we should avoid adding generations unless it's absolutely neccessary, and try to map it into the timeline or tenant ID space instead. So DLR, I think we should either change the tenant ID, or the timeline ID, and then update *that* in the control plane (as for the pageserver ID, it might be a good idea to recover on a different PS, but we don't have to put it onto a different pageserver).

Your comments, especially the request to have a single command to recreate all branches made me wonder though, maybe it is better to design the API around the whole tenant instead of just timelines? I need to think...

**Comment by arssher** [comment]
I generally agree with the RFC, but it indeed misses a significant technical question -- how to implement this attachment of timeline to pageserver which should use WAL from another (source) timeline when requesting safekeepers, but not use existing s3 data, and properly handle branching? I'm not deep enough in ps code to definitely answer that, but it could be like
- Timeline gets additional config value '(wal_source_timeline_id, LSN)' which instructs walreceiver to request different timeline id when going to safekeepers until given LSN is reached.
- API should get a way to essentially create branch by starting from initdb snapshot in s3 (+ specify the pair above). That would allow to recover timeline which never branched (root timeline).
- To recreate branch hierarchy, branch creation API call should also allow `(wal_source_timeline_id, LSN)` pair.
- Probably we should also optionally allow to create branch with future (not replayed yet) ancestor LSN so that control plane could calculate all needed branches and create them immediately without waiting for ancestor replays.

I doubt we want to replace tenant_id as well as we might want to replay only one branch.

**Comment by petuhovskiy** [comment]
Thank you for your answers, now I understand it a bit better! But I think there can be potential complications with the current branching architecture.

For example, let's say that we have the following branches:
```
main---------------->
        |
        b1-------------->
            |    |
            |    b3-------->
            |
            b2--------------->
```

- `main` – root branch, starts at LSN(0), ends at LSN(20)
- `b1` – from `main` at LSN(3), ends at LSN(20)
- `b2` – from `b1` at LSN(6), ends at LSN(20)
- `b3` – from `b1` at LSN(11), ends at LSN(20)


Currently, we store info for every branch in 3 different services:
- control-plane knows timeline_id, branch_start_lsn and branch_ancestor
- safekeeper knows timeline_id, and keeps WAL since branch_start_lsn until branch_head_lsn
- pageserver knows timeline_id, branch_start_lsn, branch_ancestor and branch_head_lsn


If we want to replay WAL for all 3 branches, what should we do? In my understanding:
1. Set maintenance mode for the tenant. Is it necessary? It means downtime for the user.
1. Recover `main` branch to the new `main'` branch
1. Recover `b1` branch to the new `b1'` branch
1. Recover `b2` branch to the new `b2'` branch
1. Recover `b3` branch to the new `b3'` branch
1. Update control-plane branches to point to the new `main'`, `b1'`, `b2'`, `b3'` branches, erasing the previous timeline_id's. Or should we create new branches instead?
1. Should we have a way to connect to the new branch and verify (amcheck?) it before lifting up the maintenance mode?
1. Deactivate maintenance mode

This looks good at first, but the internal relationships between branches become weird after we will try to continue writing WAL to them:
- All new branches will start writing at LSN(20). Will they have the real branches as the ancestors?
- Safekeepers will have WAL starting only from LSN(20) for every new timeline. Does it mean there will be mismatch, since pageserver will think that timeline starts eailier?
- What happens if pageserver dies when recovery is not finished or the result is not uploaded to S3? In this case we will need to continue recovery by reading WAL from another timeline, will pageserver persist this data in S3?
- Should we prohibit using new branches until they fully uploaded to S3?
- Is it possible to replay WAL one more time, if we will have corruption in the new branch in the future (e.g. `b3'`)? Looks complicated, since we store WAL before LSN(20) only in original branches, so we need to know where to look for this WAL. This gets more complicated if we write more WAL to new branches, and create branches from these branches.


Overall, I think we need to add more disasters examples to this RFC and have a discussion around them. Examples should have detailed steps about how to recover from this distater or explanation of why it's not possible with current tools. This will be useful to better understand implementation details and restrictions.

AFAIU, currently this RFC doesn't cover the case when internal pageserver branches mapping gets corrupted, and we need to recover the tenant by using data only from control-plane and safekeepers.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply docs/rfcs/027-pageserver-disaster-recovery.md:103]: > Personally I think it would be best to not link to the original timeline as the anchestor but instead add the capability for the timeline to be without anchestor, and handle it that way.

Note that we still need to remember timeline_id and start_lsn of the ancestor branch. This is needed to fetch WAL from safekeepers, which is stored under ancestor_timeline_id.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply docs/rfcs/027-pageserver-disaster-recovery.md:93]: > it's just the way of specifying the ID of the new timeline. alternatively, one could rename target to source and make :timeline_id be the ID of the newly created timeline. I'm fine with either, please tell me which one you like more.

I'm also fine with either, `target` is ok. But I'd rephrase description that we always create a new timeline, and `target` should not match any existing timeline. If not specified, it will be generated randomly.

File: docs/rfcs/027-pageserver-disaster-recovery.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply docs/rfcs/027-pageserver-disaster-recovery.md:84]: > the use case I can think of is a data corruption bug of the pageserver within PITR

Then we can just use existing branching API and create a new branch before the corruption.

**Comment by arpad-m** [comment]
> If we want to replay WAL for all 3 branches, what should we do? In my understanding:
> Set maintenance mode for the tenant. Is it necessary? It means downtime for the user.

One of the design constraints was that we can try out this feature with randomly chosen running tenants and checking that it works, as a way to test the feature. Therefore, we can't allow maintenance mode to be turned on.

However, for the recovery case we can assume that the user is happy that we got their data back.

> All new branches will start writing at LSN(20). Will they have the real branches as the ancestors?

I don't think that they should have the original branches as their anchestors, but instead their anchestor hierarchy should mirror the hierarchy that exists in the original. Otherwise it might be confusing. Therefore, `main'` would be the anchestor of `b1'` and `main'` would have no anchestors.

> Safekeepers will have WAL starting only from LSN(20) for every new timeline. Does it mean there will be mismatch, since pageserver will think that timeline starts eailier?

I'm not sure about the details of the protocol between safekeepers and pageservers, but from what I understand, once the pageservers have caught up with replaying the WAL to a certain point, they don't ask safekeepers for anything before that, ever. Instead they maintain their own history with delta and image layers.

Of course, there is a problem as you correctly pointed out if we want to use the restore feature on the restored branches... their WAL is partially part of another branch.

We'd probably have to store some metadata about the WAL for some parts coming from another timeline's WAL. And then consider that for for whether to delete the original timeline's WAL once we delete the original timeline.

Alternatively, and this would simplify the pageserver side a lot, we just add a feature to safekeepers to copy the WAL of timeline A into timeline B.

> AFAIU, currently this RFC doesn't cover the case when internal pageserver branches mapping gets corrupted, and we need to recover the tenant by using data only from control-plane and safekeepers.

Yeah you are right, it is not covered. This recovery method is as you have said just one of the tools in the tool belt of the engineers/support staff. I don't want perfect to be the enemy of the good, we can eventually get there if wanted :).

**Comment by petuhovskiy** [comment]
> i suppose your idea implies that the tenant ID is not being changed?

Yes, I think so. Changing tenant_id is hard because we need to maintain it in 3 different services (see above) and can be confusing for engineers debugging issues who are unaware of tenant_id change. But if it'll be the best choice, I'm not completely against changing tenant_id.

> We also want to design the feature in a way that we can use it to create a shadow replay of the WAL without affecting computes at all

It's a nice goal, but maybe we can achieve "without affecting computes" feature by other means? AFAIU we are going to use maintenance mode in control-plane, which implies downtime, so downtime is inevitable for at least some time and we may want to get back to "shadow replay" later?

> IMO we should avoid adding generations unless it's absolutely neccessary, and try to map it into the timeline or tenant ID space instead.

I have less context here, so you're probably right. But I still think it would have been nice to have a separate pageserver just for "shadow replay" purposes. For example, it can look like this:
- don't create new timelines for replay, just reset branches to start replay on the same timeline_id
- wait until we will have "generations" or other means to use more than one pageserver for the same tenant at the same time
- allocate one of the pageserver just for "shadow replay" purposes. it will allow not to interfere with any working computes and can do independent replay without affecting anything else

And if we don't use a separate pageserver, we will not be able to use the pageserver until it finishes the replay, but it's not much of an issue if we have a corrupted tenant.

**Comment by petuhovskiy** [comment]
> Therefore, we can't allow maintenance mode to be turned on.

Ok, I thought it's required from reading the notion page.

> once the pageservers have caught up with replaying the WAL to a certain point, they don't ask safekeepers for anything before that, ever. 

In most cases yes. But the general idea is that pageserver can ask WAL as many times as it wants, until pageserver uploaded data to S3 and updated `remote_consistent_lsn`.

The ps<->sk protocol can be viewed just as `read_wal(tenant_id, timeline_id, start_lsn, end_lsn)`. For now safekeepers keep WAL indefinetely and there are no restrictions.

> Alternatively, and this would simplify the pageserver side a lot, we just add a feature to safekeepers to copy the WAL of timeline A into timeline B.

It's doable. Not hard at all from the safekeepers view, the hardest part will be inside coordinator (control-plane?) to arrange copy from one set of safekeepers to another set and drive it toward completion.

>  Yeah you are right, it is not covered. This recovery method is as you have said just one of the tools in the tool belt of the engineers/support staff.

Sure. I want to understand better distinction between the issues that can be solved with this recovery and issues that are not.

**Comment by arpad-m** [comment]
I have updated the RFC to the new design.

**Review by petuhovskiy — APPROVED** [review]

**Review by jcsp — APPROVED** [review]
